### ISSUE TITLE
Numeric representation of output heatmaps (Occlusion and LayerGradCAM)

### ISSUE THREAD

**Issue by aj7989mau** [pytorch/captum#917]
I am a beginner at Python so my knowledge of terminology and useful libraries is limited. I apologise if this question has a simple answer that I have not been able to find.

I have mainly been following [this](https://pytorch.org/tutorials/beginner/introyt/captumyt.html) tutorial and working specifically with Occlusion and LayerGradCAM. I've gotten everything to work correctly, but for my project I need to do some mathematical calculations based on the value of each cell in the heat maps, which seem to be normalized to a [-1...1] range. 

I have tried printing the attribution data sent as input for the visualization, but as far as I can tell it is not a direct representation of the heat map. For example, I have tried printing both these versions of the input for Occlusion:

`attributions_occ = occlusion.attribute(input_img,
                                       target=pred_label_idx,
                                       strides=(3, 8, 8),
                                       sliding_window_shapes=(3,15, 15),
                                       baselines=0)`

and

`np.transpose(attributions_occ.squeeze().cpu().detach().numpy(), (1,2,0))`

I  have not found any information on how (or if) I could transform this input into the values that represent the heat map (normalized or not). I have not been able to find any built in method for extracting the heat map data either.

I am therefore wondering if there is any built in way to get a numerical representation of the heat map produced by Occlusion and LayerGradCAM, for example as a matrix (numpy array)? Otherwise, any indication on how to manually transform the input data to the heat map cell values would be helpful.

Thank you for any help you can give me!

**Comment by bilalsal** [comment]
Hi @aj7989mau ,

If you desire to have a 2D heatmap for the pixel importance that ignores the color channel you can try:

`attr_heatmap = np.mean(attributions_occ.squeeze().cpu().detach().numpy(), axis = 0)`

`attr_heatmap` is a  numpy array that contains the heatmap you need.

Hope this helps

**Comment by aj7989mau** [comment]
Hi @bilalsal ,

Thank you for your answer, that is very helpful and almost exactly what I need! I can most likely calculate what I need from this with a bit of work. But if you don't mind, I would love to pick your brain once more.

In the tutorial they use a 224x224 image, and the `attr_heatmap` printed with your code above matches that shape. What I specifically need is the values of the cells or sections in the heat map, which are often a larger group of pixels. Take these attribution settings for Occlusion as a simplistic example:

`attributions_occ = occlusion.attribute(input_img,
                                       target=pred_label_idx,
                                       strides=(3, 32, 32),
                                       sliding_window_shapes=(3,32, 32),
                                       baselines=0)`

This would result in a 7x7 heat map for the image, and I would like to retrieve a 7x7 matrix representing the numeric importance of each heat map cell. So it's not the indivudal pixels importance, but rather the importance of each segment in the heat map.

If you can help me with a solution for that as well I would be very grateful. If not, perhaps you can tell me if the importance value of the heat map cells is simply the mean of all the pixels they contain? So in the example above, would the importance of the top left cell of the heat map be represented by the mean of the first 32x32 pixels in the `attr_heatmap` array?

Thank you very much for your help!

**Comment by bilalsal** [comment]
Hi @aj7989mau ,

yes your understanding of the cells in the heat map is very accurate.
Here we show how you can change the sliding window parameters to obtain either a coarse or a fine heat map.
https://captum.ai/tutorials/Resnet_TorchVision_Interpret (scroll to the end).

Hope this helps,
Bilal

**Comment by aj7989mau** [comment]
Thank you very much for your help @bilalsal , I have now managed to get the data I need and gained some much needed insight regarding how the heatmap is represented. I don't know if you want me to close this topic or not so I leave it up for you to decide, but I have gotten the answers I need. 

Thank you once again for your time!
